botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import os
+import platform
+from copy import copy
+from string import ascii_letters, digits
+from typing import NamedTuple, Optional
+
+from botocore import __version__ as botocore_version
+from botocore.compat import HAS_CRT
+from botocore.utils import get_crt_version
+
+_USERAGENT_ALLOWED_CHARACTERS = ascii_letters + digits + "!$%&'*+-.^_`|~"
+_USERAGENT_ALLOWED_OS_NAMES = [
+    'windows',
+    'linux',
+    'macos',
+    'android',
+    'ios',
+    'watchos',
+    'tvos',
+    'other',
+]
+_USERAGENT_PLATFORM_NAME_MAPPINGS = {'darwin': 'macos'}
+# The name by which botocore is identified in the User-Agent header. While most
+# AWS SDKs follow a naming pattern of "aws-sdk-*", botocore and boto3 continue
+# using their existing values. Uses uppercase "B" with all other characters
+# lowercase.
+_USERAGENT_SDK_NAME = 'Botocore'
+
+
+def sanitize_user_agent_string_component(raw_str, allow_hash):
+    """Replaces all not allowed characters in the string with a dash ("-").
+
+    Allowed characters are ASCII alphanumerics and ``!$%&'*+-.^_`|~``. If
+    ``allow_hash`` is ``True``, "#"``" is also allowed.
+
+    :type allow_hash: bool
+    :param allow_hash: Whether "#" is considered an allowed character.
+    """
+    return ''.join(
+        c
+        if c in _USERAGENT_ALLOWED_CHARACTERS or (allow_hash and c == '#')
+        else '-'
+        for c in raw_str
+    )
+
+
+class UserAgentComponent(NamedTuple):
+    """
+    Component of a Botocore User-Agent header string in the standard format.
+
+    Each component consists of a prefix, a name, and a value. In the string
+    representation these are combined in the format ``prefix/name#value``.
+    """
+
+    prefix: str
+    name: str
+    value: Optional[str] = None
+
+    def to_string(self):
+        """Create string like 'prefix/name#value' from a UserAgentComponent."""
+        clean_prefix = sanitize_user_agent_string_component(
+            self.prefix, allow_hash=True
+        )
+        clean_name = sanitize_user_agent_string_component(
+            self.name, allow_hash=False
+        )
+        if self.value is None or self.value == '':
+            return f'{clean_prefix}/{clean_name}'
+        clean_value = sanitize_user_agent_string_component(
+            self.value, allow_hash=True
+        )
+        return f'{clean_prefix}/{clean_name}#{clean_value}'
```

Comment:
Two quick notes here:

1. Are these only ever called once or do we recompute per request? If these are immutable, it seems like we could cache this work. Either by doing this during initialization or some other means of memoizing the work since these are immutable.

2. We're adding `to_string` in a few places instead of `__str__`. Is there a reason we want to keep those separate or is this a Java-ism creeping in from the design?